cumulus/client/consensus/aura/src/collators/slot_based/slot_timer.rs

```diff
@@ -0,0 +1,269 @@
+	/// Client that is used for runtime calls
+	client: Arc<Client>,
+	/// Offset the current time by this duration.
+	time_offset: Duration,
+	/// Last reported core count.
+	last_reported_core_num: Option<u32>,
+	/// Slot duration of the relay chain. This is used to compute how man block-production
+	/// attempts we should trigger per relay chain block.
+	relay_slot_duration: Duration,
+	_marker: std::marker::PhantomData<(Block, Box<dyn Fn(P) + Send + Sync + 'static>)>,
+}
+
+/// Compute when to try block-authoring next.
+/// The exact time point is determined by the slot duration of relay- and parachain as
+/// well as the last observed core count. If more cores are available, we attempt to author blocks
+/// for them.
+///
+/// Returns a tuple with:
+/// - `Duration`: How long to wait until the next slot.
+/// - `Timestamp`: The timestamp to pass to the inherent
+/// - `Slot`: The AURA slot used for authoring
+fn compute_next_wake_up_time(
+	para_slot_duration: SlotDuration,
+	relay_slot_duration: Duration,
+	core_count: Option<u32>,
+	time_now: Duration,
+	time_offset: Duration,
+) -> (Duration, Timestamp, Slot) {
+	let para_slots_per_relay_block =
+		(relay_slot_duration.as_millis() / para_slot_duration.as_millis() as u128) as u32;
+	let assigned_core_num = core_count.unwrap_or(1);
+
+	// Trigger at least once per relay block, if we have for example 12 second slot duration,
+	// we should still produce two blocks if we are scheduled on every relay block.
+	let mut block_production_interval = min(para_slot_duration.as_duration(), relay_slot_duration);
+
+	if assigned_core_num > para_slots_per_relay_block &&
+		para_slot_duration.as_duration() >= relay_slot_duration
+	{
+		block_production_interval =
+			max(relay_slot_duration / assigned_core_num, BLOCK_PRODUCTION_MINIMUM_INTERVAL_MS);
+		tracing::debug!(
+			target: LOG_TARGET,
+			?block_production_interval,
+			"Expected to produce for {assigned_core_num} cores but only have {para_slots_per_relay_block} slots. Attempting to produce multiple blocks per slot."
+		);
+	}
+
+	let (duration, timestamp) =
+		time_until_next_attempt(time_now, block_production_interval, time_offset);
+	let aura_slot = Slot::from_timestamp(timestamp, para_slot_duration);
+	(duration, timestamp, aura_slot)
+}
+
+/// Returns current duration since Unix epoch.
+fn duration_now() -> Duration {
+	use std::time::SystemTime;
+	let now = SystemTime::now();
+	now.duration_since(SystemTime::UNIX_EPOCH).unwrap_or_else(|e| {
+		panic!("Current time {:?} is before Unix epoch. Something is wrong: {:?}", now, e)
+	})
+}
+
+/// Returns the duration until the next block production should be attempted.
+/// Returns:
+/// - Duration: The duration until the next attempt.
+/// - Timestamp: The time at which the attempt will take place.
+fn time_until_next_attempt(
+	now: Duration,
+	block_production_interval: Duration,
+	offset: Duration,
+) -> (Duration, Timestamp) {
+	let now = now.as_millis().saturating_sub(offset.as_millis());
```

Comment:
```suggestion
	block_production_interval: Duration,
	offset: Duration,
) -> (Duration, Timestamp) {
	let now = duration_now().as_millis().saturating_sub(offset.as_millis());
```

No real need to pass this in.